src/python/pants/core/util_rules/pants_environment.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import os
+import re
+from dataclasses import dataclass
+from typing import Dict, Optional, Sequence, cast
+
+from pants.engine.rules import side_effecting
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+@side_effecting
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class PantsEnvironment:
+    """PantsEnvironment is a representation of the environment variables the currently-executing
+    pants process was invoked with.
+
+    Since the exact contents of the environment are very specific to a given invocation of pants,
+    this type is marked as side-effecting, and can only be requested as an input to an uncacheable
+    rule. An uncacheable rule should accept this type as an input, and use `get_subset` to filter
+    the environment rules down to only the ones other rules care about as inputs, in order to avoid
+    a lot of cache invalidation if the environment changes spuriously.
+    """
+
+    env: FrozenDict[str, str]
+
+    def __init__(self, env: Optional[Dict[str, str]] = None):
+        """Initialize a `PantsEnvironment` with the current contents of the environment."""
+        if env:
+            self.env = FrozenDict(env)
+        else:
+            self.env = FrozenDict(dict(os.environ))
+
+    @classmethod
+    def create_mock(cls, mock_env: Dict[str, str]) -> "PantsEnvironment":
+        """This method is intended to be used in tests only to create a mock `PantsEnvironment`."""
+        frozen = FrozenDict(mock_env)
+        return cls(env=cast(Optional[Dict[str, str]], frozen))
+
+    def get_subset(self, requested: Sequence[str]) -> FrozenDict[str, Optional[str]]:
+        """Given a list of extra environment variable specifiers as strings, filter the contents of
+        the pants environment to only the variables that were specified as a pants configuration
+        option."""
+
+        name_value_re = re.compile(r"(\w+)=(\w+)")
+        shorthand_re = re.compile(r"(\w+)")
+
+        env_var_subset: Dict[str, Optional[str]] = {}
+
+        for env_var in requested:
+            name_value_match = name_value_re.match(env_var)
+            if name_value_match:
+                env_var_subset[name_value_match[1]] = name_value_match[2]
+            elif shorthand_re.match(env_var):
+                env_var_subset[env_var] = self.env.get(env_var)
+            else:
+                logger.warning(f"Invalid extra env var: {env_var}, skipping")
```

Comment:
Should this be an error?

Reply:
Do you mean logging.error, or a raised Exception?

I was wondering the same thing. I'm not sure what we should do. It seems likely that this PR will result in Works On My Machine issues. I could imagine an admin setting a certain value for their org, and then your laptop happening to not have that env var set. Or, maybe you set yours under a different name, and either of the two are valid. How should we approach this? I suspect logging is better than failing fast.

Reply:
My instinct is that we should generally prefer WARNING or ERROR level log messages for genuinely non-fatal errors. If it doesn't matter that this extra env variable isn't specified correctly, pants will still run. If it does matter, something else will break later on in the pants run, and there will be a conspicuous warning with the name of the incorrectly-specified variable, which is actionable.

Reply:
I'm not sure if this is more of a WARNING or ERROR log level message though, or even whether the difference between the two really matters.

Reply:
I still think this should fail eagerly, as it implies that something the user wanted to do isn't being done. And the "something else" that breaks later in the Pants run might be really subtle, e.g., it won't actually break, just give an incorrect result that the user may or may not notice. Since something is obviously wrong in this case, and we don't know how it'll affect things downstream, it's better to fail aggressively and eagerly.

Reply:
I think that's likely the correct behavior to error. Importantly, there is a remedy for the user: they can either remove the option, or they can set it.

In the error message, I recommend language like:

> You requested the environment variable `FOO` to be used during tests through the `--test-extra-environment-env` option, but `FOO` is not set in the parent process running Pants. Did you mean to set `FOO`?

Reply:
Oh, to be clear, the error I think should fail the run is if neither regex matches.

If the regex matching is fine, but there is no `FOO` set in the parent process then I think that should silently no-op. In other words, I think we should interpret `FOO` as "use FOO from the environment if it exists, otherwise don't set it". So it acts like a true allowlist.

Reply:
Bump that this should be an exception.